File: backend/app/api/v1/routes/documents.py
"""Document ingestion, semantic search, summary, and FAQ endpoints."""

from __future__ import annotations

from uuid import UUID

from fastapi import APIRouter, BackgroundTasks, Depends, File, Query, UploadFile, status
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.config import Settings, get_settings
from app.db.session import get_db_session
from app.dependencies.auth import get_verified_user
from app.dependencies.services import (
    get_document_processor,
    get_document_service,
    get_generation_service,
    get_rag_service,
)
from app.exceptions import NotFoundError
from app.models.entities import User
from app.repositories.documents import DocumentRepository
from app.schemas.common import MessageResponse, Page
from app.schemas.documents import (
    DocumentResponse,
    DocumentUploadResponse,
    FAQResponse,
    SearchRequest,
    SearchResult,
    SummaryResponse,
)
from app.services.documents import DocumentProcessor, DocumentService
from app.services.generation import DocumentGenerationService
from app.services.rag import RAGService
from app.utils.files import read_and_validate_upload

router = APIRouter()


@router.post(
    "/upload",
    response_model=DocumentUploadResponse,
    status_code=status.HTTP_202_ACCEPTED,
)
async def upload_document(
    background_tasks: BackgroundTasks,
    file: UploadFile = File(...),
    user: User = Depends(get_verified_user),
    settings: Settings = Depends(get_settings),
    service: DocumentService = Depends(get_document_service),
    processor: DocumentProcessor = Depends(get_document_processor),
) -> DocumentUploadResponse:
    """Validate, store, and asynchronously process one document."""
    validated = await read_and_validate_upload(file, max_bytes=settings.max_upload_bytes)
    document = await service.upload(user.id, validated)
    background_tasks.add_task(processor.process_one, document.id)
    return DocumentUploadResponse(document=DocumentResponse.model_validate(document))


@router.get("", response_model=Page[DocumentResponse])
async def list_documents(
    limit: int = Query(default=20, ge=1, le=100),
    offset: int = Query(default=0, ge=0),
    user: User = Depends(get_verified_user),
    session: AsyncSession = Depends(get_db_session),
) -> Page[DocumentResponse]:
    """List documents owned by the authenticated user."""
    items, total = await DocumentRepository(session).list(user.id, limit=limit, offset=offset)
    return Page(
        items=[DocumentResponse.model_validate(item) for item in items],
        total=total,
        limit=limit,
        offset=offset,
    )


@router.get("/{document_id}", response_model=DocumentResponse)
async def get_document(
    document_id: UUID,
    user: User = Depends(get_verified_user),
    session: AsyncSession = Depends(get_db_session),
) -> DocumentResponse:
    """Get one tenant-owned document."""
    document = await DocumentRepository(session).get(user.id, document_id)
    if document is None:
        raise NotFoundError("Document not found.")
    return DocumentResponse.model_validate(document)


@router.delete("/{document_id}", response_model=MessageResponse)
async def delete_document(
    document_id: UUID,
    user: User = Depends(get_verified_user),
    service: DocumentService = Depends(get_document_service),
) -> MessageResponse:
    """Delete document bytes, vectors, and visible metadata."""
    await service.delete(user.id, document_id)
    return MessageResponse(message="Document deleted.")


@router.post("/search/semantic", response_model=list[SearchResult])
async def semantic_search(
    payload: SearchRequest,
    user: User = Depends(get_verified_user),
    rag: RAGService = Depends(get_rag_service),
) -> list[SearchResult]:
    """Search across the user's selected or complete document library."""
    result = await rag.retrieve(
        user_id=user.id,
        query=payload.query,
        document_ids=payload.document_ids,
        top_k=payload.top_k,
    )
    return [SearchResult.model_validate(item) for item in result.citations]


@router.post("/{document_id}/summary", response_model=SummaryResponse)
async def summarize_document(
    document_id: UUID,
    user: User = Depends(get_verified_user),
    service: DocumentGenerationService = Depends(get_generation_service),
) -> SummaryResponse:
    """Generate summary, executive summary, and key takeaways."""
    return await service.summary(user.id, document_id)


@router.post("/{document_id}/faqs", response_model=FAQResponse)
async def generate_faqs(
    document_id: UUID,
    user: User = Depends(get_verified_user),
    service: DocumentGenerationService = Depends(get_generation_service),
) -> FAQResponse:
    """Generate exactly twenty grounded FAQs."""
    return await service.faqs(user.id, document_id)

File: backend/src/routes/documents.js
import { Router } from "express";
import multer from "multer";
import { config } from "../config.js";
import { withTransaction } from "../db.js";
import { NotFoundError } from "../errors.js";
import { requireUser } from "../middleware/auth.js";
import { integerQuery, requireUuid, validateBody } from "../middleware/validation.js";
import {
  getDocument,
  listDocuments,
  publicDocument,
} from "../repositories/documents.js";
import { searchSchema } from "../schemas.js";
import { documentProcessor, documentService } from "../services/documents.js";
import { generationService } from "../services/generation.js";
import { ragService } from "../services/rag.js";
import { validateUpload } from "../utils/files.js";

const router = Router();
const upload = multer({
  storage: multer.memoryStorage(),
  limits: { fileSize: config.maxUploadBytes, files: 1 },
});

router.use(requireUser);

router.post("/upload", upload.single("file"), async (req, res) => {
  const validated = await validateUpload(req.file, { maxBytes: config.maxUploadBytes });
  const document = await documentService.upload({
    userId: req.user.id,
    upload: validated,
    requestId: req.requestId,
  });
  res.status(202).json({
    document,
    message: "Upload accepted and document processing started.",
  });
  setImmediate(() => {
    documentProcessor.processOne(document.id).catch(() => undefined);
  });
});

router.get("/", async (req, res) => {
  const limit = integerQuery(req.query.limit, {
    defaultValue: 20,
    min: 1,
    max: 100,
    name: "limit",
  });
  const offset = integerQuery(req.query.offset, {
    defaultValue: 0,
    min: 0,
    max: Number.MAX_SAFE_INTEGER,
    name: "offset",
  });
  const page = await withTransaction(
    (client) => listDocuments(client, req.user.id, { limit, offset }),
    { userId: req.user.id },
  );
  res.json({
    items: page.items.map(publicDocument),
    total: page.total,
    limit,
    offset,
  });
});

router.post("/search/semantic", validateBody(searchSchema), async (req, res) => {
  const result = await ragService.retrieve({
    userId: req.user.id,
    query: req.validatedBody.query,
    documentIds: req.validatedBody.document_ids,
    topK: req.validatedBody.top_k,
  });
  res.json(result.citations);
});

router.get("/:documentId", async (req, res) => {
  const documentId = requireUuid(req.params.documentId, "document_id");
  const document = await withTransaction(
    (client) => getDocument(client, req.user.id, documentId),
    { userId: req.user.id },
  );
  if (!document) {
    throw new NotFoundError("Document not found.");
  }
  res.json(publicDocument(document));
});

router.delete("/:documentId", async (req, res) => {
  const documentId = requireUuid(req.params.documentId, "document_id");
  await documentService.delete({
    userId: req.user.id,
    documentId,
    requestId: req.requestId,
  });
  res.json({ message: "Document deleted." });
});

router.post("/:documentId/summary", async (req, res) => {
  res.json(
    await generationService.summary(
      req.user.id,
      requireUuid(req.params.documentId, "document_id"),
    ),
  );
});

router.post("/:documentId/faqs", async (req, res) => {
  res.json(
    await generationService.faqs(
      req.user.id,
      requireUuid(req.params.documentId, "document_id"),
    ),
  );
});

export default router;
